### ISSUE TITLE
Redirections and pipes involving blocks (like for-loops) buffer output

### ISSUE THREAD

**Issue by CGamesPlay** [fish-shell/fish-shell#1396]
When I run this script in fish 2.1.0:

```
for i in (seq 1 5); sleep 1; echo $i; end | cat
```

I should except to see a slow count to 5, one line per second. Instead, I see no output, and then, five seconds later, all 5 lines appear. This appears to be the same issue as #1302.

This appears to happen regardless of the amount of output:

```
for i in (seq 1 5); sleep 1; cat /usr/share/dict/words; end | cat
```

**Comment by enricotagliavini** [comment]
Just adding an explicit comment here in case somebody, like me, encounter the same problem and searches for already open bugs. When piping large amount of data via scripts and functions fish might exhaust available RAM memory, producing a system crash or triggering the OOM killer (on platforms having one).

One notable case: grep. If grep is defined as a function, usually to automatically add --color=auto like in bash aliases, don't process large files (producing large outputs). I triggered this by processing apache httpd log files (many GBs) and boom, OOM.

More details in issue #2803 .

Limiting the buffer size and / or forcing a flush would be nice, if possible.

**Comment by jonhoo** [comment]
This just came back to bite me again today :( In particular, I have a function that sets some secret keys in environment variables, and then ran a long-running program that produces log output. But the function causes no log output to be produced until the whole program exits, which is pretty unfortunate.

For those tracking this issue, I tried to figure out the current state of affairs: https://github.com/fish-shell/fish-shell/issues/1396#issuecomment-54508262 suggested that this would be supported with "full multithreaded execution" and references #563. That issue has now been closed as a duplicate of #238. In https://github.com/fish-shell/fish-shell/issues/238#issuecomment-286518734, @krader1961 states that there isn't a plan to add multi-threading support any time soon, so looks like we'll just have to wait patiently (or someone has to step up and write a PR).

**Comment by ObiWahn** [comment]
Until today I said that I really do not care if we use fish or bash, even though I prefer bash for my personal projects. Not being able to use `tee` to have some oupt in jenkins while logging to a file is a real deal-breaker for me.

**Comment by paddor** [comment]
I'm starting a few background jobs with redirection, like so:
```
for i in 0 1 2 3; valgrind --log-file=$i.vg ruby $i.rb > "$i.log" &; end
```
To see progress, I run `tail -f *.log`, but apparently only one of the log files is actually written to continually (line buffered, I guess). The others remain empty until the job(s) exit. Is this behavior related to this issue?

**Comment by zanchey** [comment]
No, I don't think that's related - I can get the correct behaviour with a simplified version.

**Comment by faho** [comment]
@paddor: Is `valgrind` a function?  (That includes aliases!)

`type valgrind` should tell you. If it isn't then I don't expect this to be buffered.

**Comment by paddor** [comment]
Nope, `type valgrind` tells me the path to the binary. 

Maybe Valgrind itself buffers its output if it’s not a TTY... never mind then. Thanks!

**Comment by mqudsi** [comment]
Just a ping for those subscribed: @ridiculousfish committed some improvements to this as discussed in #5635

**Comment by zanchey** [comment]
Actually all the examples given seem to work now - is there still work to be done here?

**Comment by ridiculousfish** [comment]
The way this was done was through a hack which covered the majority of cases, but not all, for example:

    for i in (seq 1 5); sleep 1; echo $i; end | begin ; cat ; end

**Comment by exploide** [comment]
I tried to fix #5340 again and this case also doesn't work, yet.

When using the following function for `ip`:
```fish
#
# Enable colored output for ip when available
#

if command -sq ip; and command ip -color=auto l >/dev/null 2>/dev/null
    function ip
        command ip -color=auto $argv
    end
end

```

`ip monitor | grep dev` does not work. (I.e. piping out of a fish function into a fish function)

**Comment by yalon** [comment]
I love using fish! This, however, is a very thorny issue, especially since aliasing commands like `grep`, which fish does by default to add coloring, cause them to behave radically different than what you'd expect coming from any other shell, and even if you do know about this issue, I think it's less convenient to start adding "command" before each part of a pipe - and even more cumbersome in case you _really want_ to execute a function as part of a pipe.
I read through this issue and #238, and I understand the appeal to add multi-threading/asyncio/ etc. concurrency models, but I also am painfully aware of how hard it is to add/change concurrency models to an existing code base.

I have another idea and before I start looking into the code base I'd love to hear from the maintainers if this can be a valid approach.

Basically, _only_ when running a pipe line, run each pipe element as a sub-shell. It's also a feature that can be done opt-in so not to break any existing behavior.

My reasoning for why this may be a reasonable approach is:
* it will make the pipe behave as a pipe for every invocation, making it intuitive (to users coming from other shells) and even more importantly, consistent
* performance overhead of a fork & exec is not an issue IMHO since I suspect the majority of the pipe use cases will run external commands anyhow (e.g. cat, grep, less, wc, ...)
* I read somewhere about sharing state between jobs using fish variables - I couldn't think of a non-contrived use case for that when piping commands, but maybe I'm not thinking of the full picture

To summarize, what about converting each pipe element to be the semi equivalent of "command fish -c <pipe element>" for each pipe element (guarded by an opt-in feature flag of course)?

**Comment by faho** [comment]
>Basically, only when running a pipe line, run each pipe element as a sub-shell. It's also a feature that can be done opt-in so not to break any existing behavior.

This is a non-starter.

Fish currently has no concept of subshells, and that's because it massively simplifies things.

E.g. imagine:

```fish
set -l lines

somecommand | while read -l line
    set lines $lines $line
end
```

If you run the `while read` in a subshell, the variable won't be shared.

This isn't a theoretical issue either, see http://mywiki.wooledge.org/BashFAQ/024.

**Comment by yalon** [comment]
Thanks @faho for the quick and detailed reply.

I didn't mean sub-shell as in some kind of IPC happening between the parent and child shell, I should have written a sub-process, just like any external command running as a pipe element.

And about the use case you wrote and the Bash FAQ link, thank you for showing a non-contrived use case :)

In my mind at least, since people coming from other shells don't expect this behavior to work, I would guess there's less fish code out there sharing mutable variables between pipe commands. On the other hand, a streaming pipe is a very common use case IMHO - dunno if there's a way to validate my claim though.

Adding to that, for input-reading use cases I think there are good alternatives like output capturing, etc.
If we agree that streaming pipes happen more often than mutating variables in pipes, then IMHO it's a good compromise to ask the user to add a something explicit where they want to mutate a state. 

This has a benefit that it makes the shell's behavior clear because you're saying a statement: you want to share the state in this step, so this step will run sequentially. 
The current status is less favorable -  you pipe some seemingly innocent commands but since a package/user/shell upgrade decided to turn that command into a function/alias, the streaming behavior changes silently. If you're piping an infinite stream for example you won't even know you have an issue.

So, here's a revised idea, again, all subject to a feature flag so not to break existing behavior:
* Each pipe element will run as a child process, same as command.
* If sharing state is desired, a new keyword or an additional symbol next to the pipe will be added - there are better people than me to suggest great syntax, but the essence of that keyword/symbol would be "this step shares the variables state of the shell may buffer the entire output of the previous command or its own output". Basically, "shared/non-streamed".
* Under this feature-flag, we can even warn/error the user if they're modifying a variable inside a pipe element without this setting this keyword/symbol first.

Thanks again for discussing this - if there's interest I can see if I can get a PR for this, but it's not a small task IMHO, so I don't want to start working on it w/o maintainers blessing.

**Comment by faho** [comment]
>In my mind at least, since people coming from other shells don't expect this behavior to work

They do, that's why it's a bash FAQ.

And it's not *just* variables either. I once had a bash script that started background processes in a `while read` block and then tried to `wait` for them after. Instead the `wait` immediately succeeded because the `while read` ran in a separate process and so the background processes weren't known to the main bash process.

>On the other hand, a streaming pipe is a very common use case IMHO 

This affects "a streaming pipe between two blocks", which is a whole lot rarer, and in most cases will just have slightly degraded performance.

In your specific case, both of these would not have buffered:

```fish
python -c 'for i in range(100000000): print(i)' | command grep 35 | grep 989
python -c 'for i in range(100000000): print(i)' | grep 35 | command grep 989
```

Most users do not hit this regularly.

>If sharing state is desired, a new keyword or an additional symbol next to the pipe will be added

This would break all code that relies on the current behavior, and there's no good way to find out which code that even is.

It's not a workable solution. I'm not saying the buffering situation shouldn't be improved, I'm saying it's not a "stop the world" sort of thing where we break everything to fix it, especially not if it's awkward to even figure out if it is.

>Thanks again for discussing this - if there's interest I can see if I can get a PR for this, 

Like I said, the multi-process thing is a non-starter, we're working on a multi-threading solution, but it's quite tricky and something that touches *a lot* of fish internals.

I do not believe this is something for an external person as their first PR.

**Comment by WesselAtWork** [comment]
Heads up.
IDK when this got fixed, but I am observing the requested functionality

```fish
><> fish -version
fish, version 3.7.0

><> for i in (seq 1 5); sleep 1; echo $i; end | cat
# 1 second of wait
1
# 1 second of wait
2
# 1 second of wait
3
# 1 second of wait
4
# 1 second of wait
5
```

**Comment by faho** [comment]
@WesselAtWork In some simple cases, fish now elides the buffer.

But if you do `| begin; cat; end` it still buffers.

**Comment by oxalica** [comment]
I did some experimental work on async script execution which fixes this issue without introducing sub-shells or fork: https://github.com/oxalica/ghoti-shell

**Comment by tgrushka** [comment]
TL;DR:
- There is no way to not buffer fish functions (the double-negative is intentional).
- Workaround A: Cannot use `alias` as a workaround because `alias` itself defines a function instead of an "alias".
- Workaround B: The UX of `abbr` is an unfriendly replacement to `alias` in the context of working around buffering of functions.

This buffering behavior is a **bug** because **piping output is a Unix power tool** that has existed since the 1970s.

Fish **fundamentally changes the behavior expectation** of piping that has unexpected and very annoying side effects **when any pipe segment involves a fish function.**

I really like the completions in fish that come out of the box, but this behavior bug is unbearably frustrating. For example, I have a justfile recipe:

```just
cosmic-logs:
    SYSTEMD_COLORS=1 journalctl -W -t cosmic-comp -f --since=now | sed -u 's/ cosmic-comp\[[0-9]*\]: / /'
```

Because `journalctl` provides no way to hide the identifier, and I don't feel like going on a scavenger hunt for custom formatting, I have to `sed` its output. OK, fine.

In order to be able to run all my `$HOME/justfile` recipes **directly** from anywhere regardless of project, I have this in my `~/.config/fish/config.fish`:

```fish
for recipe in (just -f $HOME/justfile --list 2>/dev/null | tail -n +2 | awk '{print $1}')
    eval "function $recipe; just -f $HOME/justfile $recipe \$argv; end"
end
```

So I run it like:
```
cosmic-logs
```

and get my output. Good.

But then I want to look for something:

```
cosmic-logs | grep 'error'
```

NO OUTPUT AT ALL when `error` is present in the logs.

AI suggested to use `unbuffer`, `stdbuf -o0`, `stdbuf -oL`, and all kinds of nonsense that didn't work.

So I tried:

```
function cosmic‑logs; unbuffer just ...
```

Nope.

Finally, I discovered **`fish` is causing this! Not `just`, not `sed`, not `grep`, not even `journalctl` . . . . MY SHELL.**

**Fish shell buffers the output and there is nothing in the world anyone can do within `fish` itself without writing external non-`fish` scripts to work around it.**

So I tried:
```fish
alias cosmic-logs='unbuffer just -f $HOME/justfile cosmic-logs'
```

**Should work, right? Should just run that command and allow me to pipe anything on to it. BUT NO!!!!**

**Fish wraps every `alias` in a function!!! WHAT!??!?!?!**

So then my **ONLY** workaround in fish is to:

```
abbr cosmic-logs 'unbuffer just -f $HOME/justfile cosmic-logs'
```

But then, when I type `cosmic-logs` and hit [SPACE], it **replaces it with the full command**.

The UX of requiring function expansion as the only workaround to buffered piping whenever `fish` is involved is not very nice.

And the fact that this has been open for 12 years suggests I should not hold my breath for a fix.
